### ISSUE TITLE
RequestHook regressions with native automation

### ISSUE THREAD

**Issue by Klaster1** [DevExpress/testcafe#7640]
### What is your Scenario?

All my tests run against a fake service which I redirect requests to with a `RequestHook`.

### What is the Current behavior?

1. The hook does not intercept `WebSocket` handshakes.
2. Request option changes (port, host, hostname, url) no longer affect the requests.

### What is the Expected behavior?

1. The hook intercepts `WebSocket` handshakes.
2. Request option changes (port, host, hostname, url) affect the requests.

Same as without native automation, I expect for the hook to intercept all HTTP events, including `fetch` and `WebSocket` handshakes, and successfully send requests to the web service instance assigned to the test instance.

### What is your TestCafe test code?

```
  async onRequest(e: any) {
    e.requestOptions.hostname = 'localhost';
    e.requestOptions.host = `localhost:${this.port}`;
    e.requestOptions.port = this.port.toString();
  }
```

### Steps to Reproduce

1. Add `RequestHook` that rewrites the request destination, for `fetch` requests and WebSocket handshakes.
2. Check that it works without native automation.
3. Check that it doesn't with native automation.

### TestCafe version

2.5.0

### Node.js version

18.15.0

### Command-line arguments

 testcafe chrome fixture.ts --test="Test anme" --live --native-automation

### Browser name(s) and version(s)

Chrome 112

### Platform(s) and version(s)

Windows 10 19045.2846

**Comment by Klaster1** [comment]
The app under test works with HTTP and WebSocket endpoints, I have to mock data from all sources. Mocking only HTTP requests would results in inadequate coverage.

**Comment by AlexKamaev** [comment]
@Klaster1, I was able to reproduce the RequestHook event modification issue. Let's focus on this issue in this thread for now.

We still do not clearly understand why you need to intercept WebSocket handshakes. Why do you need some code coverage for handshakes? Please share more details with us and create a separate issue for this purpose.

**Comment by AlexKamaev** [comment]
@Klaster1,

Thank you for sharing your usage scenario. Please create a separate issue about WebSockets in the Native Automation mode. Let's keep this thread only for the Request Hook redirect issue.

**Comment by Klaster1** [comment]
According to the changelog, the 2.6.0-rc1 addresses my problem, so I updated to it and run the suite again. The mocked requests are now getting blocked by something, resulting in `net::ERR_BLOCKED_BY_CLIENT` run-time error and `blocked:other` in Chrome request inspector, disabling Windows built-in firewall and antivirus did not help. Same result on Windows 11 and 10.

Another thing I noticed is that my other hook I use to mock Stripe calls doesn't work under NA too:
```
RequestMock()
  .onRequestTo({ url: /setup_intents/ })
  .respond((req, res) => {
    res.statusCode = 200;
    res.headers = { 'Access-Control-Allow-Origin': '*' };
    res.setBody(
      JSON.stringify({
        payment_method: 'pm_12345',
      }),
    );
  });
  ```

**Comment by AlexKamaev** [comment]
@Klaster1
We fixed the issue based on your description:

```JS
async onRequest(e: any) {
    e.requestOptions.hostname = 'localhost';
    e.requestOptions.host = \\`localhost:${this.port}\\`;
    e.requestOptions.port = this.port.toString();
}
```

The test code is here:
[https://github.com/DevExpress/testcafe/blob/master/test/functional/fixtures/regression/gh-7640/testcafe-fixtures/index.js](https://github.com/DevExpress/testcafe/blob/master/test/functional/fixtures/regression/gh-7640/testcafe-fixtures/index.js)
Please try running the test code on your machine.
If the test passes but your usage scenario is still not working, please modify our test or share your example.

**Comment by Klaster1** [comment]
@AlexKamaev here's a test that reproduces the Stripe hook regression. Works without native automation, and fails with.

```typescript
import { RequestMock, Selector } from 'testcafe';

fixture('Debug');

test('Debug', async (t) => {
  const hook = RequestMock()
    .onRequestTo({ url: /setup_intents/ })
    .respond((req, res) => {
      res.statusCode = 500;
      res.headers = { 'Access-Control-Allow-Origin': '*' };
      res.setBody(JSON.stringify({ error: 'Aborted by hook' }));
    });
  await t.addRequestHooks(hook);

  await t.navigateTo('https://portal.gridgain.com');
  await t.typeText(Selector('[gta="email_input"]'), 'iborisov+7640@gridgain.com');
  await t.typeText(Selector('[gta="pwd_input"]'), '1');
  await t.click(Selector('[gta="sign_in_btn"]'));
  await t.expect(Selector('[gta="create-nebula-cluster-button"]').hasAttribute('disabled')).notOk();
  await t.click(Selector('[gta="create-nebula-cluster-button"]'));
  await t.expect(Selector('dialog-provision-cluster').exists).ok();
  await t.click(Selector('.price__link'));
  await t
    .typeText(Selector(`[gta='dialog-update-credit-card__cardholder-name-input']`), 'Mighty Pirate')
    .switchToIframe(Selector(`[gta='dialog-update-credit-card__card-number-input'] iframe`))
    .typeText(Selector('input.Input'), '4242424242424242')
    .switchToMainWindow()
    .pressKey('tab')
    .switchToIframe(Selector(`[gta='dialog-update-credit-card__card-expiry-input'] iframe`))
    .typeText(Selector('input.Input'), '1030')
    .switchToMainWindow()
    .pressKey('tab')
    .switchToIframe(Selector(`[gta='dialog-update-credit-card__card-cvc-input'] iframe`))
    .typeText(Selector('input.Input'), '666')
    .switchToMainWindow()
    .pressKey('tab')
    .typeText(Selector(`[gta='dialog-update-credit-card__street-address-input']`), 'Scumm bar')
    .typeText(Selector(`[gta='dialog-update-credit-card__city-input']`), 'Melee Town')
    .typeText(Selector(`[gta='dialog-update-credit-card__region-input']`), 'Melee Island')
    .typeText(Selector(`[gta='dialog-update-credit-card__zip-input']`), '191040')
    .click(Selector(`[gta='dialog-update-credit-card__country-input']`))
    .click(Selector('mat-option').withText('Jamaica'))
    .typeText(Selector(`[gta='dialog-update-credit-card__company-input']`), 'The Pirate Company')
    .click(Selector(`[gta='dialog-update-credit-card__terms-input']`))
    .click(Selector(`[gta='dialog-update-credit-card__submit-button']`));
  await t.expect(Selector('.snackbar-error').textContent).contains('An unexpected error occurred');
});
```

As to the `net::ERR_BLOCKED_BY_CLIENT`, I was not yet able to determine if the cause lies in TestCafe or my convoluted proxy setup. Will investigate further later, at least I know a workaround for this, unlike the Stripe hook regression.

**Comment by AlexKamaev** [comment]
@Klaster1 Thank you for sharing your code. I have managed to reproduce the issue. It looks like for some reason the mock code is not called in Native Automation mode.

**Comment by miherlosev** [comment]
The example from [https://github.com/DevExpress/testcafe/issues/7640#issuecomment-1538220298](https://github.com/DevExpress/testcafe/issues/7640#issuecomment-1538220298) works correctly.
If you perform the test steps manually, you will get another error message:
![](https://github.com/DevExpress/testcafe/assets/4133518/4c4dd85b-145a-430a-8d07-957ac363f2d9)
Also, it means that in regular mode, request processing works incorrectly for this use case.
In my opinion, the best solution here is to make two expected error messages for native automation and regular modes and use the necessary one depending on the mode in use.

**Comment by Klaster1** [comment]
@miherlosev I'm not sure what you mean. Did you manage to address the discrepancy between the regular and native automation modes in the Stripe request hook above?

**Comment by AlexKamaev** [comment]
I'm still able to reproduce the described issue. I'll reopen it.

**Comment by AlexKamaev** [comment]
@Klaster1
I researched the issue with request mock of the `setup_intents` request. It seems that I found the cause. However, your website is unavailable now for some reason, so I cannot check if my fix works. Please check your usage scenario with the following testcafe package which contains the fix: [https://github.com/AlexKamaev/issues/raw/master/testcafe-2.6.1.tgz](https://github.com/AlexKamaev/issues/raw/master/testcafe-2.6.1.tgz)
Please share you results with us.

**Comment by Klaster1** [comment]
@AlexKamaev I've created a new account to use in the test, please use "iborisov+7640+2@gridgain.com" instead of "iborisov+7640@gridgain.com".

With the new build, Stripe iframes do not load at all with the NA enabled. The regular mode still works as expected.

![image](https://github.com/DevExpress/testcafe/assets/2667662/c2ee6ecc-1e80-4de5-969a-5101b19e985b)

**Comment by AlexKamaev** [comment]
@Klaster1,
I will check my fix one more time. Thank you for your cooperation.
